File: docs/algorithm/basic_algo/code/find_insert_position.py
def binary_search(lst, target):
    left, right = 0, len(lst) - 1

    while left <= right:
        mid = (left + right) // 2
        if lst[mid] == target:
            return mid
        elif lst[mid] < target:
            left = mid + 1
        else:
            right = mid - 1

    return left


nums = [1, 1, 4, 5, 7, 8, 9, 10, 11, 12]

idx = binary_search(nums, 6)
nums.insert(idx, 6)
print(nums)

idx2 = binary_search(nums, 13)
nums.insert(idx2, 13)
print(nums)

idx3 = binary_search(nums, 0)
nums.insert(idx3, 0)
print(nums)

File: docs/algorithm/basic_algo/code/find_insert_position.cpp
#include <iostream>
#include <vector>

using namespace std;

int binary_search(vector<int> &lst, int target) {
    int left = 0;
    int right = lst.size() - 1;

    while (left <= right) {
        int mid = (left + right) / 2;
        if (lst[mid] == target) {
            return mid;
        } else if (lst[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }

    return left;
}

void printVector(const vector<int> &nums) {
    for (int num: nums) {
        cout << num << " ";
    }
    cout << endl;
}

int main() {
    vector<int> nums = {1, 1, 4, 5, 7, 8, 9, 10, 11, 12};

    int idx = binary_search(nums, 6);
    nums.insert(nums.begin() + idx, 6);
    printVector(nums);

    int idx2 = binary_search(nums, 13);
    nums.insert(nums.begin() + idx2, 13);
    printVector(nums);

    int idx3 = binary_search(nums, 0);
    nums.insert(nums.begin() + idx3, 0);
    printVector(nums);

    return 0;
}